### DISCUSSION TITLE
Async migration with asyncpg

### DISCUSSION THREAD

**Discussion by redmonkez12** [sqlalchemy/alembic#1208] [Usage Questions] [answered]
Hello. I use sqlmodel with asyncpg. Is it possible to run an async migration?

```
async def upgrade() -> None:
    async with async_session() as session:
        async with session.begin():
            markets = [
                Market(name="forex"),
                Market(name="crypto"),
                Market(name="commodities"),
                Market(name="stocks"),
            ]

            session.add_all(markets)
            await session.commit()
```

**Comment by zzzeek** [comment]
what we have for asyncio is the async template , you can find docs here: https://alembic.sqlalchemy.org/en/latest/cookbook.html#using-asyncio-with-alembic

**Reply by redmonkez12** [reply]
I found it. I am fighting with it over one hour. I have no idea what I do wrong. So I decided to ask the community.

**Reply by zzzeek** [reply]
start with non-async migrations and get the feel for things first?

**Reply by redmonkez12** [reply]
Yes, the sync version works fine. I'll go with it for now.

**Comment by CaselIT** [comment, marked as answer]
You seem to be trying to use an async upgrade. That does not work.
You can do that, but you just need to use the plain session:
```py
def upgrade() -> None:
    # note the use of get_bind() here to use the same transaction used by alembic
    with Session(op.get_bind()) as session: 
            markets = [
                Market(name="forex"),
                Market(name="crypto"),
                Market(name="commodities"),
                Market(name="stocks"),
            ]

            session.add_all(markets)
            session.commit()
```

this will work even if the dialect is async and it will not block the loop.

**Reply by redmonkez12** [reply]
Nice. it works. thank you
